From stable-bounces@linux.kernel.org Mon Feb  4 22:27:28 2008
From: Oleg Nesterov <oleg@tv-sign.ru>
Date: Mon, 04 Feb 2008 22:27:18 -0800
Subject: sys_remap_file_pages: fix ->vm_file accounting
To: torvalds@linux-foundation.org
Cc: akpm@linux-foundation.org, mszeredi@suse.cz, oleg@tv-sign.ru, stable@kernel.org
Message-ID: <200802050627.m156R0c2006368@imap1.linux-foundation.org>


From: Oleg Nesterov <oleg@tv-sign.ru>

patch 8a459e44ad837018ea5c34a9efe8eb4ad27ded26 in mainline.

Fix ->vm_file accounting, mmap_region() may do do_munmap().

Signed-off-by: Oleg Nesterov <oleg@tv-sign.ru>
Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---

 mm/fremap.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/mm/fremap.c
+++ b/mm/fremap.c
@@ -190,10 +190,13 @@ asmlinkage long sys_remap_file_pages(uns
 		 */
 		if (mapping_cap_account_dirty(mapping)) {
 			unsigned long addr;
+			struct file *file = vma->vm_file;
 
 			flags &= MAP_NONBLOCK;
-			addr = mmap_region(vma->vm_file, start, size,
+			get_file(file);
+			addr = mmap_region(file, start, size,
 					flags, vma->vm_flags, pgoff, 1);
+			fput(file);
 			if (IS_ERR_VALUE(addr)) {
 				err = addr;
 			} else {
